### PR TITLE
Introduce Java binding for model-based generic tracker

### PR DESCRIPTION
**Pull request by s-trinh**
Just some tests for curiosity.
It kind of works.

ezgif com-gif-maker

I have added some comments in case of future Java developments.

### CODE REVIEW

File: modules/java/misc/mbt/gen_dict.json [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  "class_ignore_list": [
+    "vpMbDepthNormalTracker",
+    "vpMbKltTracker",
+    "vpMbScanLine",
+    "vpMbTracker",
+    "vpMbtDistanceCylinder",
+    "vpMbtMeEllipse"
+  ],
+  "ManualFuncs": {
+    "VpMbGenericTracker": {
```
Comment (s-trinh) [inline modules/java/misc/mbt/gen_dict.json]: Not able to properly bind `vpMbGenericTracker(vector_int trackerTypes)`.

Test to add all the variants in `gen_dict.json`?

```
,
      "vpMbGenericTracker" : {
        "j_code" : [
          "//",
          "// manual port",
          "// C++: vpMbGenericTracker()",
          "//",
          "//javadoc: VpMbGenericTracker::vpMbGenericTracker(const std::vector<int> &trackerTypes)\n",
          "public void vpMbGenericTracker(int[] trackerTypes)",
          "{",
          "   vpMbGenericTracker(nativeObj, trackerTypes);",
          "}"
        ],
        "jn_code" : [
          "// C++: vpMbGenericTracker(const std::vector<int> &trackerTypes)",
          "private static native void vpMbGenericTracker(long address, int[] trackerTypes);"
        ],
        "cpp_code" : [
          "//",
          "// manual port",
          "// vpMbGenericTracker(const std::vector<int> &trackerTypes)",
          "//\n",
          "JNIEXPORT jlong JNICALL Java_org_visp_mbt_VpMbGenericTracker_VpMbGenericTracker_12 (JNIEnv*, jclass, jintArray);\n",
          "JNIEXPORT jlong JNICALL Java_org_visp_mbt_VpMbGenericTracker_VpMbGenericTracker_12",
          "  (JNIEnv* env, jclass , jintArray trackerTypes)",
          "{",
          "  static const char method_name[] = \"mbt::VpMbGenericTracker_12()\";",
          "  try {",
          "   LOGD(\"%s\", method_name);",
          "   vpMbGenericTracker* _retval_ = new vpMbGenericTracker( List_to_vector_int(env, trackerTypes) );",
          "   return (jlong) _retval_;",
          "  } catch(const std::exception &e) {",
          "   throwJavaException(env, &e, method_name);",
          "  } catch (...) {",
          "   throwJavaException(env, 0, method_name);",
          "  }",
          "  return 0;",
          "}"
        ]
      }
```

File: modules/tracker/mbt/include/visp3/mbt/vpMbGenericTracker.h [hunk 2]
```diff
@@ -126,6 +126,7 @@ class VISP_EXPORT vpMbGenericTracker : public vpMbTracker
   virtual std::vector<std::string> getCameraNames() const;
 
   using vpMbTracker::getCameraParameters;
+  virtual void getCameraParameters(vpCameraParameters &camera) const;
```
Comment (s-trinh) [inline modules/tracker/mbt/include/visp3/mbt/vpMbGenericTracker.h]: It is defined in `vpMbTracker` parent class. Quick workaround is to redefine it in the `vpMbGenericTracker` class.

File: modules/tracker/mbt/src/vpMbGenericTracker.cpp [hunk 3]
```diff
@@ -1323,14 +1328,14 @@ void vpMbGenericTracker::getLline(const std::string &cameraName, std::list<vpMbt
...
-                                                                         const bool displayFullModel)
+                                                                         const vpHomogeneousMatrix &cMo_,
+                                                                         const vpCameraParameters &cam_,
+                                                                         bool displayFullModel)
 {
   std::map<std::string, TrackerWrapper *>::const_iterator it = m_mapOfTrackers.find(m_referenceCameraName);
 
   if (it != m_mapOfTrackers.end()) {
-    return it->second->getModelForDisplay(width, height, cMo, cam, displayFullModel);
+    return it->second->getModelForDisplay(width, height, cMo_, cam_, displayFullModel);
```
Comment (s-trinh) [inline modules/tracker/mbt/src/vpMbGenericTracker.cpp]: TODO:
Shadowed variable names.

File: modules/tracker/mbt/src/vpMbtXmlGenericParser.cpp [hunk 4]
```diff
@@ -1447,6 +1447,7 @@ class vpMbtXmlGenericParser::Impl
 
 vpMbtXmlGenericParser::vpMbtXmlGenericParser(int type) :  m_impl(new Impl(type))
 {
+  std::setlocale(LC_ALL, "en_US.UTF-8");
```
Comment (s-trinh) [inline modules/tracker/mbt/src/vpMbtXmlGenericParser.cpp]: Needed when calling from Java. Java defines somehow the locale to current host and PugiXML uses it to parse decimal according to the locale (issue with comma / point decimal character).

**Comment by fspindle** [comment]
@s-trinh Very nice contrib as usual

**Pull request merged**